### DISCUSSION TITLE
decoding bytearray without Unicode errors

### DISCUSSION THREAD

**Discussion by Rumidom** [micropython/micropython#14244] [Using MicroPython] [answered]
hi, how can i decode a bytearray with bytes that cant be translated, and replace those with unknown? on python 
```python
print(b'\xff\xff\xff\xff'.decode('utf8','replace'))
```
outputs: '����'

but on Micropython I still get a UnicodeError
```python
>>> print(b'\xff\xff\xff\xff'.decode('utf8','replace'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeError: 
```

**Comment by peterhinch** [comment, marked as answer]
I _think_ the code [is here](https://github.com/micropython/micropython/blob/5114f2c1ea7c05fc7ab920299967595cfc5307de/py/objstr.c#L183) - if so, the feature has not yet been implemented,

**Reply by Rumidom** [reply]
line [1946](https://github.com/micropython/micropython/blob/5114f2c1ea7c05fc7ab920299967595cfc5307de/py/objstr.c#L1946), Im still not confident/familiar with the code base to contribute myself, but I might give it a try another time if it is still to be implemented, for now i'm using a simple ASCII decoder in python:

```python
def ASCII_Decode(byte_array):
    result = ""
    for byte in byte_array:
        if 0 <= byte <= 127:
            result += chr(byte)
        else:
            result += "?"
    return result
```

**Comment by peterhinch** [comment]
Or try:
```python
byte_array = b'rats\xff foo'
bytearray([x if x <= 127 else ord('?') for x in byte_array])
```

**Reply by Rumidom** [reply]
pythonic, I like it

**Reply by peterhinch** [reply]
On second thoughts I think this is more RAM-efficient
```py
bytearray((x if x <= 127 else ord('?') for x in byte_array))
```
The first instantiates a `list` then converts it to a `bytearray`, this creates a generator.
